File: packages/uc04/src/uc04/adapters/memory/clock.py
"""Clock and id generation, injected so tests stay deterministic."""

from __future__ import annotations

from datetime import UTC, datetime, timedelta


class SystemClock:
    name = "system"

    def now(self) -> datetime:
        return datetime.now(UTC)


class FixedClock:
    """Monotonic and deterministic: one second per call."""

    name = "fixed"

    def __init__(self, base: datetime | None = None) -> None:
        self._base = base or datetime(2026, 1, 1, tzinfo=UTC)
        self._tick = 0

    def now(self) -> datetime:
        value = self._base + timedelta(seconds=self._tick)
        self._tick += 1
        return value


class SequentialIdGenerator:
    name = "sequential"

    def __init__(self) -> None:
        self._counters: dict[str, int] = {}

    def next_id(self, prefix: str) -> str:
        nxt = self._counters.get(prefix, 0) + 1
        self._counters[prefix] = nxt
        return f"{prefix}_{nxt:06d}"

File: reference/uc04-typescript/src/contracts/clock.ts
/** PORT: time & id generation, injected so tests stay deterministic. */
export interface Clock {
  nowIso(): string;
}

export interface IdGenerator {
  next(prefix: string): string;
}

export class SystemClock implements Clock {
  nowIso(): string {
    return new Date().toISOString();
  }
}

export class FixedClock implements Clock {
  private tick = 0;
  constructor(private readonly baseMs: number = Date.UTC(2026, 0, 1, 0, 0, 0)) {}
  nowIso(): string {
    // Monotonic, deterministic: one second per call.
    return new Date(this.baseMs + this.tick++ * 1000).toISOString();
  }
}

export class SequentialIdGenerator implements IdGenerator {
  private counters = new Map<string, number>();
  next(prefix: string): string {
    const n = (this.counters.get(prefix) ?? 0) + 1;
    this.counters.set(prefix, n);
    return `${prefix}_${String(n).padStart(6, '0')}`;
  }
}
